File: src/main/java/pl/laskowski/marcin/model/matrix/ArrayMatrix.java
package pl.laskowski.marcin.model.matrix;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * Created by Marcin Laskowski.
 */

public class ArrayMatrix<E> implements Matrix<E> {

    private final int sizeX;
    private final int sizeY;
    private final Object[][] data;

    public ArrayMatrix(int sizeX, int sizeY) {
        this.sizeX = sizeX;
        this.sizeY = sizeY;
        data = new Object[sizeY][sizeX];
    }

    @Override
    public int sizeX() {
        return sizeX;
    }

    @Override
    public int sizeY() {
        return sizeY;
    }

    @Override
    @SuppressWarnings("unchecked")
    public E get(int x, int y) {
        return (E) data[y][x];
    }

    @Override
    public void put(E elem, int x, int y) {
        data[y][x] = elem;
    }

    @Override
    public boolean isEmpty() {
        for (int x = 0; x < sizeX; x++) {
            for (int y = 0; y < sizeY; y++) {
                if (data[y][x] != null) return false;
            }
        }
        return true;
    }

    @Override
    public boolean contains(Object o) {
        for (int x = 0; x < sizeX; x++) {
            for (int y = 0; y < sizeY; y++) {
                if (Objects.equals(data[y][x], o)) return true;
            }
        }
        return false;
    }

    @Override
    public void forEach(BiConsumer<Integer, Integer> consumer) {
        for (int y = 0; y < sizeY; y++) {
            for (int x = 0; x < sizeX; x++) {
                consumer.accept(x, y);
            }
        }
    }

    @Override
    public void forEachX(Consumer<Integer> consumer) {
        for (int x = 0; x < sizeX; x++) {
            consumer.accept(x);
        }
    }

    @Override
    public void forEachY(Consumer<Integer> consumer) {
        for (int y = 0; y < sizeY; y++) {
            consumer.accept(y);
        }
    }

    @Override
    public boolean remove(Object o) {
        for (int x = 0; x < sizeX; x++) {
            for (int y = 0; y < sizeY; y++) {
                if (Objects.equals(data[y][x], o)) {
                    data[y][x] = null;
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public void clear() {
        for (int x = 0; x < sizeX; x++) {
            for (int y = 0; y < sizeY; y++) {
                data[y][x] = null;
            }
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ArrayMatrix<?> matrix = (ArrayMatrix<?>) o;

        if (sizeX != matrix.sizeX) return false;
        if (sizeY != matrix.sizeY) return false;
        return Arrays.deepEquals(data, matrix.data);
    }

    @Override
    public int hashCode() {
        int result = sizeX;
        result = 31 * result + sizeY;
        result = 31 * result + Arrays.deepHashCode(data);
        return result;
    }

    @Override
    public String toString() {
        return "ArrayMatrix{" +
                "sizeX=" + sizeX +
                ", sizeY=" + sizeY +
                ", data=" + Arrays.deepToString(data) +
                '}';
    }
}

File: app/src/main/java/pl/laskowski/marcin/model/matrix/ArrayMatrix.kt
package pl.laskowski.marcin.model.matrix

import java.util.function.BiConsumer
import java.util.function.Consumer

/**
 * Created by Marcin Laskowski.
 */

class ArrayMatrix<E>(
    private val sizeX: Int,
    private val sizeY: Int
) : Matrix<E> {

    private val data: Array<Array<Any?>> = Array(sizeY) { arrayOfNulls(sizeX) }

    override fun sizeX(): Int {
        return sizeX
    }

    override fun sizeY(): Int {
        return sizeY
    }

    override fun get(x: Int, y: Int): E? {
        @Suppress("UNCHECKED_CAST")
        return data[y][x] as E?
    }

    override fun put(elem: E?, x: Int, y: Int) {
        data[y][x] = elem
    }

    override val isEmpty: Boolean
        get() {
            for (x in 0 until sizeX) {
                for (y in 0 until sizeY) {
                    if (data[y][x] != null) return false
                }
            }
            return true
        }

    override operator fun contains(o: E): Boolean {
        for (x in 0 until sizeX) {
            for (y in 0 until sizeY) {
                if (data[y][x] == o) return true
            }
        }
        return false
    }

    override fun forEach(consumer: BiConsumer<Int, Int>) {
        for (y in 0 until sizeY) {
            for (x in 0 until sizeX) {
                consumer.accept(x, y)
            }
        }
    }

    override fun forEachX(consumer: Consumer<Int>) {
        for (x in 0 until sizeX) {
            consumer.accept(x)
        }
    }

    override fun forEachY(consumer: Consumer<Int>) {
        for (y in 0 until sizeY) {
            consumer.accept(y)
        }
    }

    override fun remove(o: E): Boolean {
        for (x in 0 until sizeX) {
            for (y in 0 until sizeY) {
                if (data[y][x] == o) {
                    data[y][x] = null
                    return true
                }
            }
        }
        return false
    }

    override fun clear() {
        for (x in 0 until sizeX) {
            for (y in 0 until sizeY) {
                data[y][x] = null
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val matrix = other as ArrayMatrix<*>
        if (sizeX != matrix.sizeX) return false
        return if (sizeY != matrix.sizeY) false else data.contentDeepEquals(matrix.data)
    }

    override fun hashCode(): Int {
        var result = sizeX
        result = 31 * result + sizeY
        result = 31 * result + data.contentDeepHashCode()
        return result
    }

    override fun toString(): String {
        return "ArrayMatrix{" +
                "sizeX=" + sizeX +
                ", sizeY=" + sizeY +
                ", data=" + data.contentDeepToString() +
                '}'
    }
}